main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer1.scala

```diff
@@ -0,0 +1,580 @@
+  }
+
+  private def visitEffectOp(op0: MonoAst.Op): OccurrenceAst1.Op = op0 match {
+    case MonoAst.Op(sym, spec, loc) =>
+      val sp = visitSpec(spec)
+      val fps = spec.fparams.map(visitFormalParam)
+      OccurrenceAst1.Op(sym, fps, sp, loc)
+  }
+
+  private def visitSpec(spec0: MonoAst.Spec): OccurrenceAst1.Spec = spec0 match {
+    case MonoAst.Spec(doc, ann, mod, _, functionType, retTpe, eff) =>
+      OccurrenceAst1.Spec(doc, ann, mod, functionType, retTpe, eff)
+  }
+
+  private def visitEnum(enum0: MonoAst.Enum): OccurrenceAst1.Enum = enum0 match {
+    case MonoAst.Enum(doc, ann, mod, sym, tparams, cases, loc) =>
+      val tps = tparams.map(visitTypeParam)
+      val cs = cases.map { case (sym, caze) => sym -> visitEnumCase(caze) }
+      OccurrenceAst1.Enum(doc, ann, mod, sym, tps, cs, loc)
+  }
+
+  private def visitEnumCase(case0: MonoAst.Case): OccurrenceAst1.Case = case0 match {
+    case MonoAst.Case(sym, tpes, loc) => OccurrenceAst1.Case(sym, tpes, loc)
+  }
+
+  private def visitStruct(struct0: MonoAst.Struct): OccurrenceAst1.Struct = struct0 match {
+    case MonoAst.Struct(doc, ann, mod, sym, tparams, fields, loc) =>
+      val tps = tparams.map(visitTypeParam)
+      val fs = fields.map(visitStructField)
+      OccurrenceAst1.Struct(doc, ann, mod, sym, tps, fs, loc)
+  }
+
+  private def visitStructField(field0: MonoAst.StructField): OccurrenceAst1.StructField = field0 match {
+    case MonoAst.StructField(sym, tpe, loc) =>
+      OccurrenceAst1.StructField(sym, tpe, loc)
+  }
+
+  private def visitTypeParam(tparam0: MonoAst.TypeParam): OccurrenceAst1.TypeParam = tparam0 match {
+    case MonoAst.TypeParam(name, sym, loc) => OccurrenceAst1.TypeParam(name, sym, loc)
+  }
+
+  /**
+    * Translates the given formal param `p` to the OccurrenceAst1.
+    */
+  private def visitFormalParam(fparam0: MonoAst.FormalParam): OccurrenceAst1.FormalParam = fparam0 match {
+    case MonoAst.FormalParam(sym, mod, tpe, src, loc) =>
+      OccurrenceAst1.FormalParam(sym, mod, tpe, src, loc)
+  }
+
+  /**
+    * Performs occurrence analysis on `exp00`
+    */
+  private def visitExp(sym0: Symbol.DefnSym, exp00: MonoAst.Expr): (OccurrenceAst1.Expr, OccurInfo) = {
+
+    /**
+      * Local visitor that captures `sym0` since it never changes.
+      * The implicit parameter `letBinding` is `Some(sym)` if visiting the right-hand side
+      * of a let-binding, i.e., in `let sym = exp` it recurses on `exp` with `letBinding = Some(sym)`.
+      */
+    def visit(exp0: MonoAst.Expr)(implicit letBinding: Option[VarSym]): (OccurrenceAst1.Expr, OccurInfo) = exp0 match {
+      case MonoAst.Expr.Cst(cst, tpe, loc) =>
+        (OccurrenceAst1.Expr.Cst(cst, tpe, loc), OccurInfo.One)
+
+      case MonoAst.Expr.Var(sym, tpe, loc) =>
+        (OccurrenceAst1.Expr.Var(sym, tpe, loc), OccurInfo.One + (sym -> Once))
+
+      case MonoAst.Expr.Lambda(fparam, exp, tpe, loc) =>
+        val fp = visitFormalParam(fparam)
+        val (e, o) = visit(exp)
```

Comment:
I think letbinding should be reset to none here or something?

Reply:
my thinking was that we still wouldn't want to move `???` around even though it is in a lambda.
```
let f = _ -> ???;
println(42)
```

Reply:
this program doesn't crash so moving around f is fine. and if its called then i crashes in both versions. but again i think the letbinding tracking should just be deleted